src/ripple/app/tx/impl/Clawback.cpp

```diff
@@ -0,0 +1,144 @@
+//------------------------------------------------------------------------------
+/*
+  This file is part of rippled: https://github.com/ripple/rippled
+  Copyright (c) 2023 Ripple Labs Inc.
+
+  Permission to use, copy, modify, and/or distribute this software for any
+  purpose  with  or without fee is hereby granted, provided that the above
+  copyright notice and this permission notice appear in all copies.
+
+  THE  SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+  WITH  REGARD  TO  THIS  SOFTWARE  INCLUDING  ALL  IMPLIED  WARRANTIES  OF
+  MERCHANTABILITY  AND  FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+  ANY  SPECIAL ,  DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+  WHATSOEVER  RESULTING  FROM  LOSS  OF USE, DATA OR PROFITS, WHETHER IN AN
+  ACTION  OF  CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+  OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+*/
+//==============================================================================
+
+#include <ripple/app/tx/impl/Clawback.h>
+#include <ripple/ledger/Directory.h>
+#include <ripple/ledger/View.h>
+#include <ripple/protocol/Feature.h>
+#include <ripple/protocol/Indexes.h>
+#include <ripple/protocol/Protocol.h>
+#include <ripple/protocol/TxFlags.h>
+#include <ripple/protocol/st.h>
+
+namespace ripple {
+
+NotTEC
+Clawback::preflight(PreflightContext const& ctx)
+{
+    if (!ctx.rules.enabled(featureClawback))
+        return temDISABLED;
+
+    if (auto const ret = preflight1(ctx); !isTesSuccess(ret))
+        return ret;
+
+    if (ctx.tx.getFlags() & tfClawbackMask)
+        return temINVALID_FLAG;
+
+    AccountID const issuer = ctx.tx.getAccountID(sfAccount);
+    STAmount const clawAmount(ctx.tx.getFieldAmount(sfAmount));
+
+    // The issuer field is used for the token holder instead
+    AccountID const holder = clawAmount.getIssuer();
```

Comment:
can we add a new function to the `STAmount` class which appropriate reflects the true intent of this variable? like `getHolder()` ?

I am not aware of the original intention of `getIssuer()` function name, but it is a counter-party of the intended meaning in the clawback context

Reply:
Such functionality doesn't belong in the `STAmount` class which is _already_ overburdened with cruft.

Reply:
@ckeshava The intent of this is written in the [spec](https://github.com/XRPLF/XRPL-Standards/tree/master/XLS-39d-clawback).

> The sub-field issuer within Amount represents the token holder's address instead of the issuer's.

We are just using the issuer field because it would save us from the need to create one extra parameter. But I agree it's confusing to some people.

Reply:
yes, I remember seeing it in the spec. I felt it would be easier to understand if it was the other way around.

Reply:
I'm not quite sure if there is other way around this, without adding an additional transaction parameter.

There are two accounts involved in a clawback transaction - the issuer and the holder. The `Account` field is the issuer, who would always be the one submitting the transaction. 

The holder's account is specified in the Amount field like this
```
{
    "currency": "USD",
    "value": "153.75",
    "issuer": "r9cZA1mLK5R5Am25ArfXFmqgNwjZgnfk59"
}
```
Here, and the `issuer` sub-field has the address`r9cZA1mLK5R5Am25ArfXFmqgNwjZgnfk59` which is the holder's address. So after this `Amount` field is serialized into `STAmount` object, calling `getIssuer()` on this object will return the holder's address. This is definitely an exception.

Reply:
Not really, `issuer` is the right naming for `Amount`. `clawback` is merely using this field in a "altered" way.

There is more info in the doc about [Token Amount](https://xrpl.org/currency-formats.html#token-amounts) .
![image](https://github.com/XRPLF/rippled/assets/35279399/51e316ed-94a4-4167-ae3a-aab574d38e8e)